### PR TITLE
Add JSON Support to V2 Engine (#217)

### PR DESCRIPTION
**Pull request by GumpacG**
### Description
Phase 1 of adding JSON Support to V2 Engine
For example: The following request was handled by V2 engine: 
```
curl --location 'http://localhost:9200/_plugins/_sql?format=json' --header 'Content-Type: application/json' --data '{ 
    "query": "SELECT * FROM calcs LIMIT 1;"
}'
{
    "took": 13,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 17,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": "calcs",
                "_id": "1",
                "_score": 1.0,
                "_source": {
                    "int0": 1,
                    "bool3": true,
                    "time1": "19:36:22",
                    "bool2": false,
                    "int2": 5,
                    "int1": -3,
                    "str3": "e",
                    "int3": 8,
                    "str1": "CLAMP ON LAMPS",
                    "str2": "one",
                    "time0": "1899-12-30T21:07:32Z",
                    "num1": 8.42,
                    "datetime0": "2004-07-09T10:17:35Z",
                    "num0": 12.3,
                    "datetime1": null,
                    "num4": null,
                    "key": "key00",
                    "num3": -11.52,
                    "bool1": true,
                    "num2": 17.86,
                    "bool0": true,
                    "str0": "FURNITURE",
                    "date3": "1986-03-20",
                    "date2": "1977-04-20",
                    "date1": "2004-04-01",
                    "date0": "2004-04-15",
                    "zzz": "a"
                }
            }
        ]
    }
}
```
 
Returning the raw response from OpenSearch made IT tests fail due to missing in-memory operations, difference in aggregation type (V2 is composite_bucket), and some bugs were discovered. The solution for these failing IT tests were to fall back to legacy for all functions, expect failures for IT tests with a comment for issues related.

**List of expected failures in legacy IT**:

- DateFormatIT.and [(1436)](https://github.com/opensearch-project/sql/issues/1436)

- DateFormatIT.equalTo [(1436)](https://github.com/opensearch-project/sql/issues/1436)

- DateFormatIT.greaterThan [(1436)](https://github.com/opensearch-project/sql/issues/1436) 

- DateFormatIT.greaterThanOrEqualTo [(1436)](https://github.com/opensearch-project/sql/issues/1436)

- DateFormatIT.lessThan [(1436)](https://github.com/opensearch-project/sql/issues/1436)

- DateFormatIT.lessThanOrEqualTo [(1436)](https://github.com/opensearch-project/sql/issues/1436)

- DateFormatIT.or [(1436)](https://github.com/opensearch-project/sql/issues/1436)
- DateFormatIT.sortByDateFormat [(1436)](https://github.com/opensearch-project/sql/issues/1436)
- QueryFunctionsIT.wildcardQuery (Dependent on [PR #1314](https://github.com/opensearch-project/sql/pull/1314))
- QueryIT.selectAllWithFieldAndOrderBy [(785)](https://github.com/opensearch-project/sql/issues/785) 

- QueryIT.selectAllWithFieldReturnsAll [(785)](https://github.com/opensearch-project/sql/issues/785) 

- QueryIT.selectAllWithFieldReverseOrder [(785)](https://github.com/opensearch-project/sql/issues/785)

- QueryIT.selectAllWithMultipleFields [(785)](https://github.com/opensearch-project/sql/issues/785) 

- QueryIT.notLikeTests [(1425)](https://github.com/opensearch-project/sql/issues/1425) 

### Proposal for phase 2 of JSON support for V2
https://github.com/opensearch-project/sql/issues/1450

### Issues Resolved
https://github.com/opensearch-project/sql/issues/1317
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java [hunk 1]
```diff
@@ -80,7 +83,7 @@ public void open() {
     request = requestBuilder.build();
     iterator = Collections.emptyIterator();
     queryCount = 0;
-    fetchNextBatch();
+    rawResponse = fetchNextBatch().getRawResponse();
```
Comment (MaxKsyunz) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java:86]: There's a problem here.
 `fetchNextBatch` is called in `hasNext` when it's piecing together a scroll request. Scroll request is issued when limit clause specifies number larger than index window size. 
 
 Raw responses for each batch need to be added together to get raw response for the whole query.

Here's an integration test that demonstrates the problem. It gets count of rows in a table, runs the same SQL query in JDBC and JSON and confirms that count is the expected value (1000), JDBC response 1000 entries, and that JSON and JDBC responses have the same number of entries. That last assert fails -- JSON response only has 3 entries -- the  window value for this test.

```java
  @Test
  public void limitLargerThanWindow() throws IOException {
    setMaxResultWindow(TestsConstants.TEST_INDEX_ACCOUNT, 3);

    var count = executeJdbcRequest("SELECT COUNT(*) FROM " + TestsConstants.TEST_INDEX_ACCOUNT);
    var query = "SELECT * FROM " + TestsConstants.TEST_INDEX_ACCOUNT + " LIMIT 1000";
    var requestBody = String.format("{\"query\" : \" %s\"}", query);
    var r = executeJdbcRequest(query);

    // The call below throws an exception
    var response = query(requestBody);
    var hits = response.getHits();
    assertEquals(1000, count.getJSONArray("datarows").getJSONArray(0).getInt(0));
    assertEquals(1000, r.getJSONArray("datarows").length());
    assertEquals(r.getJSONArray("datarows").length(), hits.length);
  }
```

**Review by MaxKsyunz — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ * This visitor's sole purpose is to throw UnsupportedOperationExceptions
+ * for unsupported features in the new engine when JSON format is specified.
+ * Unsupported features in V2 are ones the produce results that differ from
+ * legacy results.
+ */
+public class JsonSupportVisitor extends AbstractNodeVisitor<Void, JsonSupportVisitorContext> {
+  @Override
+  public Void visit(Node node, JsonSupportVisitorContext context) {
+    visitChildren(node, context);
+    return null;
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java:28]: The Boolean return type is a good change. 
I'm not convinced that removing the Exception throws in this instance is the right way.  Try...Catch blocks are slightly more expensive in processing time, but then we have the satisfaction of knowing that the NodeVisitor is interrupted.

**Review by GabeFernandez310 — APPROVED** [review]

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java (hunk 1 above)
Comment (GumpacG) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java:86]: Legacy does not support having `max_result_window` less than limit size and throws the exception:
```
"error": {
        "root_cause": [
            {
                "type": "illegal_argument_exception",
                "reason": "Result window is too large, from + size must be less than or equal to: [3] but was [1000]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."
            }
        ]
```

In order to have parity with legacy, the plugin will also fall back to legacy when `LIMIT` is included in the query.

File: core/src/main/java/org/opensearch/sql/planner/physical/PhysicalPlan.java [hunk 3]
```diff
@@ -45,4 +45,9 @@ public ExecutionEngine.Schema schema() {
     throw new IllegalStateException(String.format("[BUG] schema can been only applied to "
         + "ProjectOperator, instead of %s", toString()));
   }
+
+  public String getRawResponse() {
+    return getChild().stream().map(PhysicalPlan::getRawResponse)
+        .filter(r -> r != null && !r.isEmpty()).findFirst().orElse("");
```
[the remark this answers is not in the document]
Comment (GumpacG) [reply core/src/main/java/org/opensearch/sql/planner/physical/PhysicalPlan.java]: Updated in [4e692fe](https://github.com/opensearch-project/sql/pull/1464/commits/4e692fef7864c719a3a09870e309d745deb50098)

File: core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (GumpacG) [reply core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java]: Updated in [4e692fe](https://github.com/opensearch-project/sql/pull/1464/commits/4e692fef7864c719a3a09870e309d745deb50098)

**Review by MaxKsyunz — APPROVED** [review]

**Review by Yury-Fridlyand — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import org.opensearch.sql.ast.tree.Limit;
+import org.opensearch.sql.ast.tree.Project;
+
+/**
+ * This visitor's sole purpose is to throw UnsupportedOperationExceptions
+ * for unsupported features in the new engine when JSON format is specified.
+ * Unsupported features in V2 are ones the produce results that differ from
+ * legacy results.
+ */
+public class JsonSupportVisitor extends AbstractNodeVisitor<Boolean, JsonSupportVisitorContext> {
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java:25]: I'm thinking if any easy way to do this. As I understand, JSON format only makes sense if the physical query plan for a query only has a Project (item is regular field) and OpenSearchIndexScan. If so, can we do this validation later?

**Review by dai-chen — COMMENTED** [review]
I have a high level question about the implementation which I asked earlier on PoC PR too: 

For now this JSON format only makes senses for OpenSearch storage. However, special logic are added in many places, `QueryService`, `PhysicalPlan`, `QueryResponse`, `ExecutionEngine` etc. Many of them are very fundamental and on critical code path of our core engine. Can we think about if any more extensible way:

1) Either make this JSON format and code changes work for all other storage;
2) Or minimize the special logic, ex. use `_raw` meta field and only use a special JSON response formatter to work with it.

Thanks!

**Comment by Yury-Fridlyand** [comment]
This PR is a first phase of JSON support implementation. Second phase assumes support of JSON response format for all datasources and queries. We have a PoC on [test-json-pagination](https://github.com/Bit-Quill/opensearch-project-sql/tree/test-json-pagination) branch, it keep no raw response from OpenSearch/Prometheus/whatever, but it generates a response which looks 1:1 like returned by OpenSearch with resultset of the SQL plugin.
See example:
```json
$ curl -s -XPOST http://localhost:9200/_plugins/_sql?format=json -H 'Content-Type: application/json' -d '{"query": "SELECT int0 + 1, 2, now() FROM calcs limit 2"}' | jq
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": null,
        "_id": null,
        "_score": 0,
        "fields": {
          "2": 2,
          "int0 + 1": 2,
          "now()": "2023-03-17 13:07:47"
        }
      },
      {
        "_index": null,
        "_id": null,
        "_score": 0,
        "fields": {
          "2": 2,
          "int0 + 1": null,
          "now()": "2023-03-17 13:07:47"
        }
      }
    ]
  }
}
```
It is not completed yet (`_id`, `_index` fields are empty - depends on #1456, `took` is always 0), but it works fine. For another datasource, if fills meta info (e.g. `_shards`) with default values.
Once we will complete the fix and resolve all dependencies, we will publish Phase 2 implementation of JSON response format.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java [hunk 5]
```diff
@@ -49,7 +49,8 @@ public void execute(PhysicalPlan physicalPlan, ExecutionContext context,
               result.add(plan.next());
             }
 
-            QueryResponse response = new QueryResponse(physicalPlan.schema(), result);
+            String rawResponse = physicalPlan.getRawResponse();
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/executor/OpenSearchExecutionEngine.java:52]: If it is json format, result is not been used, only rawResponse will be used in response, right? Does it means the query plan should be different, if the query plan to fetch raw data?

File: core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java (hunk 4 above)
Comment (penghuo) [reply core/src/main/java/org/opensearch/sql/analysis/JsonSupportVisitor.java:25]: +1, As I understand it, all unsupported cases will still fallback to the V1 engine, so there won't be a significant difference, right?

**Review by penghuo — COMMENTED** [review]
Regarding the long-term plan, do we intend to fully deprecate the legacy engine? I believe that the legacy engine excels in supporting the JSON format, correct? What we do eventually is translate SQL statement to DSL as much as possible.

**Comment by penghuo** [comment]
> This PR is a first phase of JSON support implementation. Second phase assumes support of JSON response format for all datasources and queries. We have a PoC on [test-json-pagination](https://github.com/Bit-Quill/opensearch-project-sql/tree/test-json-pagination) branch, it keep no raw response from OpenSearch/Prometheus/whatever, but it generates a response which looks 1:1 like returned by OpenSearch with resultset of the SQL plugin. See example:
> 
> ```json
> $ curl -s -XPOST http://localhost:9200/_plugins/_sql?format=json -H 'Content-Type: application/json' -d '{"query": "SELECT int0 + 1, 2, now() FROM calcs limit 2"}' | jq
> {
>   "took": 0,
>   "timed_out": false,
>   "_shards": {
>     "total": 1,
>     "successful": 1,
>     "skipped": 0,
>     "failed": 0
>   },
>   "hits": {
>     "total": {
>       "value": 2,
>       "relation": "eq"
>     },
>     "max_score": 1,
>     "hits": [
>       {
>         "_index": null,
>         "_id": null,
>         "_score": 0,
>         "fields": {
>           "2": 2,
>           "int0 + 1": 2,
>           "now()": "2023-03-17 13:07:47"
>         }
>       },
>       {
>         "_index": null,
>         "_id": null,
>         "_score": 0,
>         "fields": {
>           "2": 2,
>           "int0 + 1": null,
>           "now()": "2023-03-17 13:07:47"
>         }
>       }
>     ]
>   }
> }
> ```
> 
> It is not completed yet (`_id`, `_index` fields are empty - depends on #1456, `took` is always 0), but it works fine. For another datasource, if fills meta info (e.g. `_shards`) with default values. Once we will complete the fix and resolve all dependencies, we will publish Phase 2 implementation of JSON response format.
> 
> **UPD** Phase 2 is tracked in #1450

I believe that what we are targeting here is specific to the OpenSearch raw data response. Do other data sources have similar requirements?

**Comment by Yury-Fridlyand** [comment]
> I believe that what we are targeting here is specific to the OpenSearch raw data response. Do other data sources have similar requirements?

Not yet, but you're welcome to define them in #1450. It is possible to generate OpenSearch-like JSON using result set from Prometheus. Or create OpenSearch response for OpenSearch queries and Prometheus-like for Prometheus.

**Comment by dai-chen** [comment]
> > I believe that what we are targeting here is specific to the OpenSearch raw data response. Do other data sources have similar requirements?
> 
> Not yet, but you're welcome to define them in #1450. It is possible to generate OpenSearch-like JSON using result set from Prometheus. Or create OpenSearch response for OpenSearch queries and Prometheus-like for Prometheus.

IMO, if we touch core logic as what we're doing in this PR, we should clearly define what's the meaning of JSON format for all data sources. Otherwise, we may want to minimize such special logic that only work for OpenSearch.

@acarbonetto @Yury-Fridlyand 
I expressed the similar concern in meta field PR: https://github.com/opensearch-project/sql/pull/1456#discussion_r1151006697. We've put lots of efforts to separate modules and abstraction since the beginning of V2 engine. It's true that OpenSearch is the primary storage and we're prioritizing legacy engine deprecate. However, if we keep adding special OpenSearch logic to core, I'm afraid V2 engine would be hard to maintain as legacy in future and all our previous efforts would be in vain.

**Pull request closed without merging**